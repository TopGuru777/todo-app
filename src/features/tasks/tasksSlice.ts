import { createSlice, PayloadAction } from '@reduxjs/toolkit';

type TaskStatus = 'Pending' | 'In Progress' | 'Completed';

export interface Task {
  id: string;
  title: string;
  description: string;
  status: TaskStatus;
  dueDate: string;
}

interface TasksState {
  tasks: Task[];
}

const initialState: TasksState = {
  tasks: [],
};

const tasksSlice = createSlice({
  name: 'tasks',
  initialState,
  reducers: {
    addTask: (state, action: PayloadAction<Task>) => {
      state.tasks.push(action.payload);
    },
    editTask: (state, action: PayloadAction<Task>) => {
      const index = state.tasks.findIndex(task => task.id === action.payload.id);
      if (index !== -1) {
        state.tasks[index] = action.payload;
      }
    },
    deleteTask: (state, action: PayloadAction<string>) => {
      state.tasks = state.tasks.filter(task => task.id !== action.payload);
    },
    updateTasks: (state, action: PayloadAction<Task[]>) => {
      state.tasks = action.payload;
    },
  },
});

export const { addTask, editTask, deleteTask, updateTasks } = tasksSlice.actions;
export default tasksSlice.reducer;