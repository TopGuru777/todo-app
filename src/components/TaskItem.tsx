import React, { useState, useCallback } from 'react';
import { useDispatch } from 'react-redux';
import Modal from 'react-modal';
import { editTask, deleteTask } from '../features/tasks/tasksSlice';
import { toast } from 'react-toastify';

interface TaskItemProps {
  task: {
    id: string;
    title: string;
    description: string;
    status: 'Pending' | 'In Progress' | 'Completed';
    dueDate: string;
  };
}

const TaskItem: React.FC<TaskItemProps> = React.memo(({ task }) => {
  const dispatch = useDispatch();
  const [isModalOpen, setIsModalOpen] = useState(false);
  const [isEditing, setIsEditing] = useState(false);
  const [title, setTitle] = useState(task.title);
  const [description, setDescription] = useState(task.description);
  const [dueDate, setDueDate] = useState(task.dueDate);
  const [status, setStatus] = useState(task.status);

  const handleSave = useCallback(() => {
    dispatch(editTask({ ...task, title, description, dueDate, status }));
    setIsEditing(false);
    toast.success('Task updated successfully!');
  }, [dispatch, task, title, description, dueDate, status]);

  const handleCancel = useCallback(() => {
    setTitle(task.title);
    setDescription(task.description);
    setDueDate(task.dueDate);
    setStatus(task.status);
    setIsEditing(false);
  }, [task]);

  const handleDelete = useCallback(() => {
    dispatch(deleteTask(task.id));
    setIsModalOpen(false);
    toast.error('Task deleted!');
  }, [dispatch, task.id]);

  const statusBadgeColor = {
    'Pending': 'bg-yellow-500',
    'In Progress': 'bg-blue-500',
    'Completed': 'bg-green-500',
  };

  return (
    <>
      <div
        className="bg-white shadow-md rounded p-4 mb-4 cursor-pointer flex justify-between items-center"
        onClick={() => setIsModalOpen(true)}
      >
        <div>
          <h3 className="text-lg font-bold truncate w-full">{task.title}</h3>
          <p className="truncate w-full text-sm">{task.description}</p>
          <p className={`text-sm ${new Date(task.dueDate) < new Date() ? 'text-red-500' : ''}`}>
            Due: {task.dueDate}
          </p>
        </div>
        <span className={`text-white text-xs font-semibold px-2 py-1 rounded ${statusBadgeColor[task.status]}`}>
          {task.status}
        </span>
      </div>

      <Modal
        isOpen={isModalOpen}
        onRequestClose={() => setIsModalOpen(false)}
        contentLabel="Task Details"
        className="bg-white p-6 rounded-lg shadow-xl max-w-lg mx-auto mt-24 relative"
        overlayClassName="fixed inset-0 bg-black bg-opacity-50 flex justify-center items-center"
      >
        {!isEditing ? (
          <>
            <h2 className="text-2xl font-bold mb-4">{title}</h2>
            <p className="mb-2">{description}</p>
            <p className="mb-4 text-sm">Due: {dueDate}</p>
            <span className={`text-white text-xs font-semibold px-2 py-1 rounded ${statusBadgeColor[status]}`}>
              {status}
            </span>
            <div className="mt-6">
              <button
                onClick={() => setIsEditing(true)}
                className="bg-blue-600 hover:bg-blue-700 text-white font-semibold py-2 px-4 rounded-lg shadow-md transition duration-300 mr-2"
              >
                Edit
              </button>
              <button
                onClick={() => setIsModalOpen(false)}
                className="bg-gray-600 hover:bg-gray-700 text-white font-semibold py-2 px-4 rounded-lg shadow-md transition duration-300"
              >
                Close
              </button>
            </div>
          </>
        ) : (
          <>
            <h2 className="text-2xl font-bold mb-4">Edit Task</h2>
            <input
              type="text"
              value={title}
              onChange={(e) => setTitle(e.target.value)}
              className="shadow appearance-none border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline mb-2"
              placeholder="Task Title"
            />
            <textarea
              value={description}
              onChange={(e) => setDescription(e.target.value)}
              className="shadow appearance-none border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline mb-2"
              placeholder="Task Description"
            />
            <input
              type="date"
              value={dueDate}
              onChange={(e) => setDueDate(e.target.value)}
              className="shadow appearance-none border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline mb-2"
            />
            <select
              value={status}
              onChange={(e) => setStatus(e.target.value as 'Pending' | 'In Progress' | 'Completed')}
              className="shadow appearance-none border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline mb-4"
            >
              <option value="Pending">Pending</option>
              <option value="In Progress">In Progress</option>
              <option value="Completed">Completed</option>
            </select>
            <div className="mt-6">
              <button
                onClick={handleSave}
                className="bg-green-600 hover:bg-green-700 text-white font-semibold py-2 px-4 rounded-lg shadow-md transition duration-300 mr-2"
              >
                Save
              </button>
              <button
                onClick={handleCancel}
                className="bg-gray-600 hover:bg-gray-700 text-white font-semibold py-2 px-4 rounded-lg shadow-md transition duration-300 mr-2"
              >
                Cancel
              </button>
              <button
                onClick={handleDelete}
                className="mt-4 bg-red-600 hover:bg-red-700 text-white font-semibold py-2 px-4 rounded-lg shadow-md transition duration-300"
              >
                Delete
              </button>
            </div>
          </>
        )}
      </Modal>
    </>
  );
});

export default TaskItem;