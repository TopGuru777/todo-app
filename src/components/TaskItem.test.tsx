import { render, fireEvent, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import TaskItem from './TaskItem';
import { editTask, deleteTask, Task } from '../features/tasks/tasksSlice';
import { ToastContainer } from 'react-toastify';

const mockStore = configureStore([]);
jest.mock('../features/tasks/tasksSlice', () => ({
  editTask: jest.fn(),
  deleteTask: jest.fn(),
}));

describe('TaskItem', () => {
  let store: ReturnType<typeof mockStore>;
  const task: Task = {
    id: '1',
    title: 'Test Task',
    description: 'Test Description',
    status: 'Pending',
    dueDate: '2024-11-08',
  };

  beforeEach(() => {
    store = mockStore({});
    store.dispatch = jest.fn();
  });

  test('edits a task', () => {
    render(
      <Provider store={store}>
        <TaskItem task={task} />
        <ToastContainer />
      </Provider>
    );

    fireEvent.click(screen.getByText(/Test Task/i));
    fireEvent.click(screen.getByText(/Edit/i));

    fireEvent.change(screen.getByPlaceholderText(/Task Title/i), {
      target: { value: 'Updated Task' },
    });
    fireEvent.change(screen.getByPlaceholderText(/Task Description/i), {
      target: { value: 'Updated Description' },
    });
    fireEvent.click(screen.getByText(/Save/i));

    expect(store.dispatch).toHaveBeenCalledWith(
      editTask({
        ...task,
        title: 'Updated Task',
        description: 'Updated Description',
      })
    );
  });

  test('deletes a task', () => {
    render(
      <Provider store={store}>
        <TaskItem task={task} />
        <ToastContainer />
      </Provider>
    );

    fireEvent.click(screen.getByText(/Test Task/i));
    fireEvent.click(screen.getByText(/Edit/i));
    fireEvent.click(screen.getByText(/Delete/i));

    expect(store.dispatch).toHaveBeenCalledWith(deleteTask(task.id));
  });

  test('matches snapshot', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <TaskItem task={task} />
      </Provider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});