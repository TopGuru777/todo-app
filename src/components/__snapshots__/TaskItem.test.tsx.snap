// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskItem matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="bg-white shadow-md rounded p-4 mb-4 cursor-pointer flex justify-between items-center"
  >
    <div>
      <h3
        class="text-lg font-bold truncate w-full"
      >
        Test Task
      </h3>
      <p
        class="truncate w-full text-sm"
      >
        Test Description
      </p>
      <p
        class="text-sm text-red-500"
      >
        Due: 2024-11-08
      </p>
    </div>
    <span
      class="text-white text-xs font-semibold px-2 py-1 rounded bg-yellow-500"
    >
      Pending
    </span>
  </div>
</DocumentFragment>
`;
